torchmultimodal/models/gpt.py

```diff
@@ -0,0 +1,405 @@
+        attn_mask (Tensor, optional): Tensor of dimension ``(b, out_seq_len, in_seq_len)``. Contains 1s for
+            positions to attend to and 0s for masked positions. Defaults to ``None``.
+        head_mask (Tensor, optional): Tensor of dimension ``(b, h, out_seq_len, in_seq_len)``. Contains 1s
+            for attention heads to use and 0s for masked heads. Defaults to ``None``.
+        use_cache (bool, optional): If ``True``, caches past key/value tensors for faster decoding. If ``False``,
+            recomputes key and value for each decoding step. Defaults to ``False``.
+        causal (bool. optional): If ``True``, use causal attention. Defaults to ``False``.
+        return_attn_weights (bool, optional): If ``True``, returns attention probabilities of each transformer
+            layer. Defaults to ``False``.
+        return_hidden_states (bool): If ``True``, returns the embeddings of each transformer layer. Defaults to
+            ``False``.
+    """
+
+    def __init__(
+        self,
+        decoder_layer: nn.Module,
+        num_layers: int = 12,
+    ) -> None:
+        super().__init__()
+        self.layers = get_clones(decoder_layer, num_layers)
+        self.num_layers = num_layers
+
+    def forward(
+        self,
+        hidden_states: Tensor,
+        attn_mask: Optional[Tensor] = None,
+        head_mask: Optional[Tensor] = None,
+        use_cache: bool = False,
+        causal: bool = False,
+        return_attn_weights: bool = False,
+        return_hidden_states: bool = False,
+    ) -> TransformerDecoderOutput:
+        all_hidden_states: Tuple[Tensor, ...] = () if return_hidden_states else None
+        all_attentions: Tuple[Tensor, ...] = () if return_attn_weights else None
+        all_past_key_values: Tuple[Dict[str, Tensor], ...] = () if use_cache else None
+
+        for layer in self.layers:
+            if return_hidden_states:
+                all_hidden_states = all_hidden_states + (hidden_states,)
+            layer_outputs = layer(
+                hidden_states,
+                attn_mask,
+                head_mask,
+                use_cache,
+                causal,
+                return_attn_weights,
+            )
+            hidden_states = layer_outputs.hidden_states
+            if return_attn_weights:
+                all_attentions = all_attentions + (layer_outputs.attention_weights,)
+            if use_cache:
+                all_past_key_values = all_past_key_values + (
+                    layer_outputs.past_key_values,
+                )
+
+        if return_hidden_states:
+            all_hidden_states = all_hidden_states + (hidden_states,)
+
+        return TransformerDecoderOutput(
+            last_hidden_states=hidden_states,
+            hidden_states=all_hidden_states,
+            attention_weights=all_attentions,
+            past_key_values=all_past_key_values,
+        )
+
+
+class SiLU(nn.Module):
```

Comment:
We also have this in CLIP now so might be worth consolidating: https://github.com/facebookresearch/multimodal/blob/1c4fa802684a87a26ca203ef18bf3110ab10ff1a/torchmultimodal/modules/encoders/clip_text_encoder.py#L14

Reply:
I noticed that but not sure what our stance here since we are gradually shifting towards model centric components.
And it will be weird for `gpt.py` to take a dep on `clip.py` -- could be confusing to users.
Or we can create `torchmultimodal/modules/layers/activation.py` and move it over there?

Reply:
Yeah agree that we should not take a dep on CLIP here. I like the option of adding activation.py but fine to leave as-is for now if you prefer

Reply:
Ok. Created a task to move to activation.py as a follow-up.